sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UTF8StringBuilder.java

```diff
@@ -45,34 +47,29 @@ private void grow(int neededSize) {
           "exceeds size limitation " + ARRAY_MAX);
     }
     final int length = totalSize() + neededSize;
-    if (buffer.length < length) {
+    if (buffer.size() < length) {
       int newLength = length < ARRAY_MAX / 2 ? length * 2 : ARRAY_MAX;
-      final byte[] tmp = new byte[newLength];
-      Platform.copyMemory(
-        buffer,
-        Platform.BYTE_ARRAY_OFFSET,
-        tmp,
-        Platform.BYTE_ARRAY_OFFSET,
-        totalSize());
+      final ByteArrayMemoryBlock tmp = new ByteArrayMemoryBlock(newLength);
+      MemoryBlock.copyMemory(buffer, tmp, totalSize());
       buffer = tmp;
     }
   }
 
   private int totalSize() {
-    return cursor - Platform.BYTE_ARRAY_OFFSET;
+    return length;
```

Comment:
Since `totalSize` is privately used only, and now it is simply `length`, I think we can remove it and just use `length`.